tests/src/Kernel/Entity/GetBundleByBundleTest.php

```diff
@@ -0,0 +1,557 @@
+
+  /**
+   * {@inheritdoc}
+   */
+  public static $modules = [
+    'block_content',
+    'field',
+    'node',
+    'og',
+    'system',
+    'user',
+  ];
+
+  /**
+   * Test groups.
+   *
+   * @var \Drupal\Core\Entity\EntityInterface[][]
+   */
+  protected $groups = [];
+
+  /**
+   * Test group content.
+   *
+   * @var \Drupal\Core\Entity\EntityInterface
+   */
+  protected $groupContent;
+
+  /**
+   * The group manager.
+   *
+   * @var \Drupal\og\GroupManager
+   */
+  protected $groupManager;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    parent::setUp();
+
+    $this->installConfig(['og']);
+    $this->installEntitySchema('block_content');
+    $this->installEntitySchema('node');
+    $this->installEntitySchema('og_membership');
+    $this->installEntitySchema('user');
+    $this->installSchema('system', 'sequences');
+
+    $this->groupManager = $this->container->get('og.group.manager');
+
+    // Create four groups of two different entity types.
+    for ($i = 0; $i < 2; $i++) {
+      $bundle = "group_$i";
+      NodeType::create([
+        'name' => $this->randomString(),
+        'type' => $bundle,
+      ])->save();
+      Og::groupManager()->addGroup('node', $bundle);
+
+      BlockContentType::create(['id' => $bundle])->save();
+      Og::groupManager()->addGroup('block_content', $bundle);
+    }
+  }
+
+  /**
+   * Tests retrieval of bundles that are referenc[ed|ing] bundles.
+   *
+   * This tests the retrieval of the relations between groups and group content
+   * and vice versa. The retrieval of groups that are referenced by group
+   * content is done by GroupManager::getGroupBundleIdsByGroupContenBundle()
+   * while GroupManager::getGroupContentBundleIdsByGroupBundle() handles the
+   * opposite case.
+   *
+   * Both methods are tested here in a single test since they are very similar,
+   * and not having to set up the entire relationship structure twice reduces
+   * the total test running time.
+   *
+   * @param array $relationships
+   *   An array indicating the relationships between groups and group content
+   *   bundles that need to be set up in the test.
+   * @param array $expected_group_by_group_content
+   *   An array containing the expected results for the call to
+   *   getGroupBundleIdsByGroupContentBundle().
+   * @param array $expected_group_content_by_group
+   *   An array containing the expected results for the 4 calls to
+   *   getGroupContentBundleIdsByGroupBundle() that will be made in the test.
+   *
+   * @covers ::getGroupBundleIdsByGroupContentBundle
+   * @covers ::getGroupContentBundleIdsByGroupBundle
+   *
+   * @dataProvider getBundleIdsByBundleProvider
+   */
+  public function testGetBundleIdsByBundle(array $relationships, array $expected_group_by_group_content, array $expected_group_content_by_group) {
+    // Set up the relations as indicated in the test.
+    foreach ($relationships as $group_content_entity_type_id => $group_content_bundle_ids) {
```

Comment:
it's a little hard following the test, I'm going to review this part tomorrow when brain is a bit more fresh :)